### PR TITLE
Search Builder fails with an error when searching for State if the location type differs from the display name.

### PR DESCRIPTION
**Pull request by kainuk**
Overview
----------------------------------------
Search Builder fails with an error when searching for State if the location type differs from the display name.

To reproduce
- Goto https://dmaster.demo.civicrm.org/civicrm/admin/locationType?reset=1
- Change the name of Other (nr 4) to other.
- Start the search builder.
- Create the following search 
screen01
The search ends with a DB Error - Syntax Error

Before
----------------------------------------
To reproduce
- Goto https://dmaster.demo.civicrm.org/civicrm/admin/locationType?reset=1
- Change the name of Other (nr 4) to other.
- Start the search builder.
- Create the following search 
screen01
The search ends with a DB Error - Syntax Error

After
----------------------------------------
The same as above except that the search now gives a result.

Technical Details
----------------------------------------
The cause of the problem is that locationtypes can be retrieved on two different ways in CiviCRM e.g.
`CRM_Core_DAO_Address::buildOptions('location_type_id', 'validate');`
and 
` CRM_Core_PseudoConstant::get('CRM_Core_DAO_Address', 'location_type_id')`

When name and display_name have a different case these functions return different results.

My solution is to replace the second with the first.

A test is added to the PR. It fails with unknown variable. I do not know how to catch it

Comments
----------------------------------------
Also on gitlab (https://lab.civicrm.org/dev/core/issues/607).
In the test database name and display are the same. The effect is that this kind of bug is not catched. Tip for implementers. Take care that name and display name are always the same.

### CODE REVIEW

File: CRM/Contact/BAO/Query.php [hunk 1]
```diff
@@ -2130,7 +2130,7 @@ public function restWhere(&$values) {
 
     $setTables = TRUE;
 
-    $locationType = CRM_Core_PseudoConstant::get('CRM_Core_DAO_Address', 'location_type_id');
+    $locationType = CRM_Core_DAO_Address::buildOptions('location_type_id', 'validate');
     if (isset($locType[1]) && is_numeric($locType[1])) {
```
Comment (mattwire) [inline CRM/Contact/BAO/Query.php]: The same thing could be achieved with:
`CRM_Core_PseudoConstant::get('CRM_Core_DAO_Address', 'location_type_id', [], 'validate');`

I'm not sure which  of `CRM_Core_PseudoConstant::get` and CRM_Core_DAO_Address::buildOptions` is preferred - @eileenmcnaughton ??

In this case, I think you could further simplify this to something like:
`CRM_Core_PseudoConstant::getName('CRM_Core_DAO_Address', 'location_type_id', $locType[1]);`

Comment (kainuk) [reply CRM/Contact/BAO/Query.php]: Maybe `buildOptions` is not the best solution, but I chose it to be the same as the code used on line 2790.

Comment (monishdeb) [reply CRM/Contact/BAO/Query.php]: @mattwire I think its ok to use buildOptions in here as we are doing in other places too:
```
CRM//Contact/Selector.php:419:        $locationTypes = CRM_Core_DAO_Address::buildOptions('location_type_id', 'validate');
CRM//Contact/BAO/Query.php:1016:    $locationTypes = CRM_Core_DAO_Address::buildOptions('location_type_id', 'validate');
CRM//Contact/BAO/Query.php:2371:      $locationType = CRM_Core_DAO_Address::buildOptions('location_type_id', 'validate');
CRM//Contact/BAO/Query.php:2790:          $locationTypes = CRM_Core_DAO_Address::buildOptions('location_type_id', 'validate');
CRM//Core/BAO/Mapping.php:1132:    $locationTypes = CRM_Core_DAO_Address::buildOptions('location_type_id', 'validate');
```

**Comment by monishdeb** [comment]
The patch successfully fixes the issue. Here's the screencasts:

BEFORE
----------
before1

AFTER
--------
after

**Comment by monishdeb** [comment]
Merging on basis of:
* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) __PASS__
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __PASS__
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __PASS__
    * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __PASS__
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __PASS__
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __PASS__
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __PASS__
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __PASS__

**Pull request merged**